google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2761,4 +2790,77 @@ private void initMetricsCollection(MetricRegistry metricRegistry, List<LabelValu
         // TODO: Remove metric.
         ignored -> 0L);
   }
+
+  private void initOpenTelemetryMetricsCollection(
+      OpenTelemetry openTelemetry, Attributes attributes) {
+
+    if (openTelemetry != null) {
+      Meter meter = openTelemetry.getMeter(MetricRegistryConstants.Scope);
+
+      meter
+          .gaugeBuilder(MAX_ALLOWED_SESSIONS_NAME)
+          .setDescription(MAX_ALLOWED_SESSIONS_DESCRIPTION)
+          .setUnit(COUNT)
+          .buildWithCallback(
+              measurement -> {
+                measurement.record(options.getMaxSessions(), attributes);
+              });
+
+      meter
+          .gaugeBuilder(MAX_IN_USE_SESSIONS_NAME)
+          .setDescription(MAX_IN_USE_SESSIONS_DESCRIPTION)
+          .setUnit(COUNT)
+          .buildWithCallback(
+              measurement -> {
+                measurement.record(this.maxSessionsInUse, attributes);
+              });
+
+      AttributesBuilder attributesBuilder;
+      if (attributes != null) {
+        attributesBuilder = attributes.toBuilder();
+      } else {
+        attributesBuilder = Attributes.builder();
+      }
+      Attributes attributesInUseSessions =
+          attributesBuilder.put(SESSIONS_TYPE, NUM_SESSIONS_IN_USE_NAME).build();
+      Attributes attributesAvailableSessions =
+          attributesBuilder.put(SESSIONS_TYPE, NUM_SESSIONS_AVAILABLE_NAME).build();
+      meter
+          .upDownCounterBuilder(NUM_SESSIONS_IN_POOL_NAME)
+          .setDescription(NUM_SESSIONS_IN_POOL_DESCRIPTION)
+          .setUnit(COUNT)
+          .buildWithCallback(
+              measurement -> {
+                measurement.record(this.numSessionsInUse, attributesInUseSessions);
```

Comment:
Access to properties like `this.numSessionsInUse` and `this.sessions` is normally guarded with a `synchronized (lock)` block to prevent concurrent reads/writes to take place. Do we need that for this as well? (When are these callbacks being called?)
And if we were to add `synchronized (lock)` in these callbacks: Do the threads that call these callbacks take any other locks as well? If yes, can we guarantee that there won't be a deadlock?

Reply:
The frequency of callback will depend on what exporter customers will be using with what export interval. 

For instance cloudmonitoringexporter will be pushing data every 10 sec as default.

Reply:
I'm not necessarily worried about the frequency, more about potential race conditions, and maybe also not so much about these specific cases. But we also call `sessions.size()` for example without taking a lock. `sessions.size()` is undefined if it is called while other threads are modifying the collection.